main/src/library/Vector.flix

```diff
@@ -0,0 +1,1389 @@
+    /// searching from left to right.
+    ///
+    pub def indexOfLeft(a: a, v: Vector[a]): Option[Int32] with Eq[a] =
+        def loop(i) = {
+            if (i >= length(v))
+                -1
+            else if (get(i, v) == a)
+                i
+            else
+                loop(i + 1)
+        };
+        let i = loop(0);
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Optionally returns the position of the first occurrence of `a` in `v`
+    /// searching from right to left.
+    ///
+    pub def indexOfRight(a: a, v: Vector[a]): Option[Int32] with Eq[a] =
+        def loop(i) = {
+            if (i < 0)
+                -1
+            else if (get(i, v) == a)
+                i
+            else
+                loop(i - 1)
+        };
+        let i = loop(length(v) - 1);
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Return the positions of the all the occurrences of `a` in `v`.
+    ///
+    pub def indices(a: a, v: Vector[a]): Vector[Int32] with Eq[a] =
+        findIndices(b -> a == b, v)
+
+    ///
+    /// Alias for `findLeft`.
+    ///
+    pub def find(f: a -> Bool \ ef, v: Vector[a]): Option[a] \ ef =
+        findLeft(f, v)
+
+    ///
+    /// Optionally returns the first element of `v` that satisfies the predicate `f` when searching from left to right.
+    ///
+    pub def findLeft(f: a -> Bool \ ef, v: Vector[a]): Option[a] \ ef =
+        match findIndexOfLeft(f, v) {
+            case None    => None
+            case Some(i) => Some(get(i, v))
+        }
+
+    ///
+    /// Optionally returns the first element of `v` that satisfies the predicate `f` when searching from right to left.
+    ///
+    pub def findRight(f: a -> Bool \ ef, v: Vector[a]): Option[a] \ ef =
+        match findIndexOfRight(x -> f(x), v) {
+            case None    => None
+            case Some(i) => Some(get(i, v))
+        }
+
+    ///
+    /// Returns a vector of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns an empty vector if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Vector[Int32] =
+        let len = e - b;
+        init(ix -> b + ix, len)
+
+    ///
+    /// Returns a vector with the element `x` repeated `n` times.
+    ///
+    /// Returns an empty vector if `n <= 0`.
+    ///
+    pub def repeat(n: Int32, x: a): Vector[a] =
+        init(_ -> x, n)
+
+    ///
+    /// Alias for `scanLeft`.
+    ///
+    pub def scan(f: (b, a) -> b \ ef, s: b, v: Vector[a]): Vector[b] \ ef =
+        scanLeft(f, s, v)
+
+    ///
+    /// Accumulates the result of applying `f` to `v` going left to right.
+    ///
+    /// That is, the result is of the form: `[s , f(s, x1), f(f(s, x1), x2),  ...]`.
+    ///
+    pub def scanLeft(f: (b, a) -> b \ ef, s: b, v: Vector[a]): Vector[b] \ ef = region rc {
+        let acc = ref s @ rc;
+        buildVector(rc, length(v) + 1, arr -> {
+            Array.put(s, 0, arr);
```

Comment:
There uses of put are still problematic because we will not `Vector.put`.

One option could be to scanLeft into an array and then copy it to a Vector.

Reply:
`scanLeft` and `scanRight` now initialize an interim array directly instead of using `buildVector` so they should be clearer as to how they work.